contrib/pg_upgrade/pg_upgrade.c

```
@@ -405,6 +405,10 @@ setup(char *argv0, bool *live_check)
 	/* no postmasters should be running, except for a live check */
 	if (pid_lock_file_exists(old_cluster.pgdata))
 	{
+		if (canConnectToServer(&old_cluster) && user_opts.check)
```

This change in logic will allow servers with stale PID files to live on? That doesn't seem like an improvement. What is the problem we're trying to solve?

---

Huh, didn't know that was a problem. The problem we're trying to solve is stated in the description.

---

I don't understand the description, it's not an error condition that the target cluster is down and the source cluster up, so not showing an error seems reasonable. Or am I missing something?

---

At the moment a "*failure*" message gets printed when the source cluster is up because of `start_postmaster` failing to actually start. We want to avoid printing the "*failure*" message during check while keeping the source cluster up.

---

Aha, so you're saying that the user sees a failure message from starting the postmaster when it's already up and we enter live-check mode? I'm a bit thick today and misread the intent here.

I still feel that changing the logic here is the wrong way to go, the comment below makes that very clear. When executing the command in `exec_proc` via `start_postmaster`, the caller can suppress error reporting with the `report_error` flag which seems like the better approach here since we mainly want to fix a UI issue.

---

I was attempting to not change the signature of `start_postmaster` to help reduce diffs.

---

This was actually fixed in upstream in commit d25ee30031b08ad1348a090914c2af6bc640a832, so cherry-picking that is probably the easiest option while also reducing the risk of conflicts.